addons/sale_coupon/models/sale_coupon_program.py

```diff
@@ -118,10 +118,10 @@ def unlink(self):
 
     def toggle_active(self):
         super(SaleCouponProgram, self).toggle_active()
-        if not self.active:
-            coupons = self.filtered(lambda p: p.promo_code_usage == 'code_needed').mapped('coupon_ids')
+        for program in self.filtered(lambda x: not x.active):
+            coupons = program.filtered(lambda p: p.promo_code_usage == 'code_needed').mapped('coupon_ids')
             coupons.filtered(lambda x: x.state != 'used').write({'state': 'expired'})
```

Comment:
```python
if program.promo_code_usage == 'code_needed':
  program.coupon_ids.filtered(...).state = 'expired'
```

The filtered is useless since program should only contain one record.

Reply:
Actually would even be better to merge both `filtered()`, and to to it in 11.0, not in 13.0, code should be 100% identical.

```python
for program in self.filtered(lambda p: not p.active and p.promo_code_usage == 'code_needed'):
    program.mapped('coupon_ids').filtered(lambda c: c.state != 'used').write({'state': 'expired'})
```
A one liner is also probably better, if linted correctly, to avoid doing multiple write when only one is possible.
```python
self.filtered(
    lambda p: not p.active and p.promo_code_usage == 'code_needed'
).mapped('coupon_ids').filtered(
    lambda c: c.state != 'used'
).write({'state': 'expired'})
```
or
```python
self.filtered(lambda p: not p.active and p.promo_code_usage == 'code_needed') \
    .mapped('coupon_ids').filtered(lambda c: c.state != 'used').write({'state': 'expired'})
```
Might do it, got to test which is linted correctly.

Reply:
Well, the for loop could even be dropped, and we could write on the filtered result.
```python
programs = self.filtered(lambda p: not p.active and p.promo_code_usage == 'code_needed')
programs.mapped('coupon_ids').write({'state': 'expired'})
programs.mapped('discount_line_product_id').write({'active': False}
```

Reply:
> Well, the for loop could even be dropped, and we could write on the filtered result.

Yes, that's exactly what I said in my previous message.

But the code you provided is wrong. `toggle_active` is not only used to set `active` to `False` but also to `True`, `.write({'active': False}` is not correct.
Also, the filtering stuff is only to write on `coupon_ids`, not on `discount_line_product_id`.
This part is wrong too, as `programs` are now always filtered on inactive ones, thus it won't reactivate `discount_line_product_id` when unarchiving programs.

Correct fix, I think, would be:
```python
def toggle_active(self):
    super(SaleCouponProgram, self).toggle_active()
    for program in self:
        program.discount_line_product_id.active = program.active
    coupons = self.filtered(lambda p: not p.active and p.promo_code_usage == 'code_needed').mapped('coupon_ids')
    coupons.filtered(lambda x: x.state != 'used').write({'state': 'expired'})
```
This will do the `write` in one request, which is the most important, and readability seems good enough.

Reply:
My bad, I did indeed read your post too fast :man_facepalming:
(and my discount_line_product_id part was wrong :c)

Last proposed solution lgtm :+1:.